File: ios/OpenEOSCore/Sources/OpenEOSCore/CCAPIMultipartLiveView.swift
import Foundation

enum CCAPIMultipartLiveView {
    static let maximumFrameBytes = 12 * 1024 * 1024
    private static let maximumBoundaryCharacters = 200

    static func boundary(from contentType: String?) throws -> String {
        let parts = (contentType ?? "").split(separator: ";", omittingEmptySubsequences: false)
        guard parts.first?.trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased() == "multipart/x-mixed-replace" else {
            throw CCAPIError.invalidResponse(
                "Canon multipart Live View returned \(contentType ?? "no Content-Type"); expected multipart/x-mixed-replace."
            )
        }
        let rawBoundary = parts.dropFirst().compactMap { part -> String? in
            let components = part.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
            guard components.count == 2,
                  components[0].trimmingCharacters(in: .whitespacesAndNewlines).lowercased() == "boundary" else {
                return nil
            }
            return components[1].trimmingCharacters(in: .whitespacesAndNewlines)
        }.first ?? ""
        let unquoted = rawBoundary.hasPrefix("\"") && rawBoundary.hasSuffix("\"") && rawBoundary.count >= 2
            ? String(rawBoundary.dropFirst().dropLast())
            : rawBoundary
        let boundary = unquoted.hasPrefix("--") ? String(unquoted.dropFirst(2)) : unquoted
        guard !boundary.isEmpty,
              boundary.count <= maximumBoundaryCharacters,
              boundary.unicodeScalars.allSatisfy({ (33...126).contains($0.value) && $0.value != 34 }) else {
            throw CCAPIError.invalidResponse(
                "Canon multipart Live View returned a missing or invalid ASCII boundary."
            )
        }
        return boundary
    }
}

struct CCAPIMultipartParser {
    private static let maximumLineBytes = 8 * 1024
    private static let maximumHeaderBytes = 16 * 1024
    private static let maximumHeaderCount = 32

    private enum State {
        case boundary(scanned: Int)
        case headers(values: [String: String], bytes: Int)
        case body(length: Int)
        case closed
    }

    private let delimiter: String
    private let closingDelimiter: String
    private var state = State.boundary(scanned: 0)
    private var buffer = Data()

    init(boundary: String) {
        delimiter = "--\(boundary)"
        closingDelimiter = "--\(boundary)--"
    }

    mutating func append(_ data: Data) throws -> [Data] {
        buffer.append(data)
        guard buffer.count <= CCAPIMultipartLiveView.maximumFrameBytes + Self.maximumHeaderBytes else {
            throw CCAPIError.invalidResponse("Canon multipart Live View buffer exceeded the safety limit.")
        }
        var frames: [Data] = []
        parsing: while true {
            switch state {
            case let .boundary(scanned):
                guard let line = try readLine() else { break parsing }
                let total = scanned + line.utf8.count
                guard total <= Self.maximumHeaderBytes else {
                    throw CCAPIError.invalidResponse("Canon multipart Live View preamble exceeded the safety limit.")
                }
                if line == delimiter {
                    state = .headers(values: [:], bytes: 0)
                } else if line == closingDelimiter {
                    state = .closed
                } else {
                    state = .boundary(scanned: total)
                }
            case let .headers(values, bytes):
                guard let line = try readLine() else { break parsing }
                if line.isEmpty {
                    let contentType = values["content-type"]?.split(separator: ";", maxSplits: 1).first?
                        .trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
                    guard contentType == "image/jpeg" else {
                        throw CCAPIError.invalidResponse("Canon multipart Live View part is not image/jpeg.")
                    }
                    let lengthText = values["content-length"] ?? ""
                    guard !lengthText.isEmpty,
                          lengthText.unicodeScalars.allSatisfy({ (48...57).contains($0.value) }),
                          let length = Int(lengthText),
                          (1...CCAPIMultipartLiveView.maximumFrameBytes).contains(length) else {
                        throw CCAPIError.invalidResponse(
                            "Canon multipart Live View returned an invalid or unsafe Content-Length."
                        )
                    }
                    state = .body(length: length)
                    continue
                }
                let total = bytes + line.utf8.count
                guard total <= Self.maximumHeaderBytes, values.count < Self.maximumHeaderCount,
                      let separator = line.firstIndex(of: ":") else {
                    throw CCAPIError.invalidResponse(
                        "Canon multipart Live View returned malformed or oversized part headers."
                    )
                }
                let name = String(line[..<separator])
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                    .lowercased()
                let value = String(line[line.index(after: separator)...])
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                guard !name.isEmpty, values[name] == nil else {
                    throw CCAPIError.invalidResponse("Canon multipart Live View returned a duplicate part header.")
                }
                var updated = values
                updated[name] = value
                state = .headers(values: updated, bytes: total)
            case let .body(length):
                guard buffer.count >= length else { break parsing }
                let frame = Data(buffer.prefix(length))
                buffer.removeFirst(length)
                guard frame.count >= 4,
                      frame.prefix(2) == Data([0xFF, 0xD8]),
                      frame.suffix(2) == Data([0xFF, 0xD9]) else {
                    throw CCAPIError.invalidResponse(
                        "Canon multipart Live View part did not contain a complete JPEG frame."
                    )
                }
                frames.append(frame)
                state = .boundary(scanned: 0)
            case .closed:
                guard buffer.allSatisfy({ $0 == 0x0D || $0 == 0x0A }) else {
                    throw CCAPIError.invalidResponse("Canon multipart Live View returned data after its closing boundary.")
                }
                buffer.removeAll(keepingCapacity: false)
                break parsing
            }
        }
        return frames
    }

    mutating func finish() throws {
        _ = try append(Data())
        guard case .closed = state, buffer.isEmpty else {
            throw CCAPIError.invalidResponse("Canon multipart Live View ended before a complete closing boundary.")
        }
    }

    private mutating func readLine() throws -> String? {
        guard let newline = buffer.firstIndex(of: 0x0A) else {
            guard buffer.count <= Self.maximumLineBytes else {
                throw CCAPIError.invalidResponse("Canon multipart Live View line exceeded the safety limit.")
            }
            return nil
        }
        let count = buffer.distance(from: buffer.startIndex, to: newline)
        guard count <= Self.maximumLineBytes else {
            throw CCAPIError.invalidResponse("Canon multipart Live View line exceeded the safety limit.")
        }
        var line = Data(buffer.prefix(count))
        buffer.removeFirst(count + 1)
        if line.last == 0x0D { line.removeLast() }
        return String(decoding: line, as: Unicode.UTF8.self)
    }
}

final class CCAPIMultipartLiveViewSession: @unchecked Sendable {
    private var iterator: AsyncThrowingStream<Data, Error>.Iterator
    private let continuation: AsyncThrowingStream<Data, Error>.Continuation
    private let worker: Task<Void, Never>
    private let response: CameraHTTPStreamResponse

    init(response: CameraHTTPStreamResponse, boundary: String) {
        self.response = response
        let output = AsyncThrowingStream<Data, Error>.makeStream(bufferingPolicy: .bufferingNewest(1))
        iterator = output.stream.makeAsyncIterator()
        continuation = output.continuation
        worker = Task {
            var parser = CCAPIMultipartParser(boundary: boundary)
            do {
                for try await chunk in response.chunks {
                    try Task.checkCancellation()
                    for frame in try parser.append(chunk) {
                        output.continuation.yield(frame)
                    }
                }
                try parser.finish()
                output.continuation.finish()
            } catch {
                output.continuation.finish(throwing: error)
            }
        }
    }

    func nextFrame() async throws -> Data {
        guard let frame = try await iterator.next() else {
            throw CCAPIError.invalidResponse("Canon multipart Live View stream ended before the next frame.")
        }
        return frame
    }

    func close() {
        worker.cancel()
        response.cancel()
        continuation.finish(throwing: CancellationError())
    }
}

File: android/app/src/main/java/dev/openeos/control/data/CcapiMultipartLiveView.kt
package dev.openeos.control.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Call
import okhttp3.Response
import java.io.Closeable
import java.io.InputStream
import java.nio.charset.StandardCharsets
import java.util.Locale
import java.util.concurrent.TimeUnit

internal suspend fun closeCcapiMultipartSession(session: AutoCloseable?) {
    if (session == null) return
    withContext(Dispatchers.IO) { session.close() }
}

internal const val CCAPI_MULTIPART_MAX_FRAME_BYTES = 12 * 1024 * 1024
private const val MAX_BOUNDARY_CHARS = 200
private const val MAX_LINE_BYTES = 8 * 1024
private const val MAX_HEADER_BYTES = 16 * 1024
private const val MAX_HEADER_COUNT = 32

internal fun parseCcapiMultipartBoundary(contentType: String?): String {
    val value = contentType?.trim().orEmpty()
    val components = value.split(';')
    require(components.firstOrNull()?.trim()?.equals("multipart/x-mixed-replace", ignoreCase = true) == true) {
        "Canon multipart Live View returned ${contentType ?: "no Content-Type"}; expected multipart/x-mixed-replace."
    }
    val rawBoundary = components.drop(1).firstNotNullOfOrNull { parameter ->
        val name = parameter.substringBefore('=', missingDelimiterValue = "").trim()
        if (!name.equals("boundary", ignoreCase = true)) return@firstNotNullOfOrNull null
        parameter.substringAfter('=', missingDelimiterValue = "").trim()
    }.orEmpty()
    val boundary = if (rawBoundary.length >= 2 && rawBoundary.first() == '"' && rawBoundary.last() == '"') {
        rawBoundary.substring(1, rawBoundary.lastIndex)
    } else {
        rawBoundary
    }.removePrefix("--")
    require(boundary.isNotBlank() && boundary.length <= MAX_BOUNDARY_CHARS) {
        "Canon multipart Live View returned a missing or invalid boundary."
    }
    require(boundary.all { it.code in 33..126 && it != '"' }) {
        "Canon multipart Live View returned a non-ASCII boundary."
    }
    return boundary
}

internal class CcapiMultipartFrameReader(
    private val input: InputStream,
    boundary: String,
) {
    private val delimiter = "--$boundary"
    private val closingDelimiter = "$delimiter--"

    fun nextFrame(): ByteArray? {
        when (seekBoundary()) {
            Boundary.NEXT -> Unit
            Boundary.CLOSED -> return null
        }

        val headers = linkedMapOf<String, String>()
        var headerBytes = 0
        while (true) {
            val line = readLine() ?: error("Canon multipart Live View ended while reading part headers.")
            if (line.isEmpty()) break
            headerBytes += line.toByteArray(StandardCharsets.ISO_8859_1).size
            require(headerBytes <= MAX_HEADER_BYTES && headers.size < MAX_HEADER_COUNT) {
                "Canon multipart Live View part headers exceeded the safety limit."
            }
            val separator = line.indexOf(':')
            require(separator > 0) { "Canon multipart Live View returned a malformed part header." }
            val name = line.substring(0, separator).trim().lowercase(Locale.ROOT)
            val value = line.substring(separator + 1).trim()
            require(name !in headers) { "Canon multipart Live View returned a duplicate $name header." }
            headers[name] = value
        }

        require(headers["content-type"]?.substringBefore(';')?.trim()
            ?.equals("image/jpeg", ignoreCase = true) == true) {
            "Canon multipart Live View part is not image/jpeg."
        }
        val lengthText = headers["content-length"]
            ?: error("Canon multipart Live View part is missing Content-Length.")
        require(lengthText.isNotEmpty() && lengthText.all(Char::isDigit)) {
            "Canon multipart Live View returned an invalid Content-Length."
        }
        val length = lengthText.toLongOrNull()
            ?: error("Canon multipart Live View returned an invalid Content-Length.")
        require(length in 1..CCAPI_MULTIPART_MAX_FRAME_BYTES.toLong()) {
            "Canon multipart Live View frame length $length is outside the safety limit."
        }
        val frame = ByteArray(length.toInt())
        var offset = 0
        while (offset < frame.size) {
            val count = input.read(frame, offset, frame.size - offset)
            if (count < 0) error("Canon multipart Live View ended before the JPEG frame was complete.")
            offset += count
        }
        require(
            frame.size >= 4 &&
                frame[0] == 0xFF.toByte() && frame[1] == 0xD8.toByte() &&
                frame[frame.lastIndex - 1] == 0xFF.toByte() && frame[frame.lastIndex] == 0xD9.toByte()
        ) { "Canon multipart Live View part did not contain a complete JPEG frame." }
        return frame
    }

    private fun seekBoundary(): Boundary {
        var scanned = 0
        while (true) {
            val line = readLine() ?: error("Canon multipart Live View ended before the next boundary.")
            scanned += line.length
            require(scanned <= MAX_HEADER_BYTES) {
                "Canon multipart Live View preamble exceeded the safety limit."
            }
            when (line) {
                delimiter -> return Boundary.NEXT
                closingDelimiter -> return Boundary.CLOSED
            }
        }
    }

    private fun readLine(): String? {
        val bytes = ArrayList<Byte>()
        while (true) {
            val value = input.read()
            if (value < 0) {
                if (bytes.isEmpty()) return null
                break
            }
            if (value == '\n'.code) break
            require(bytes.size < MAX_LINE_BYTES) { "Canon multipart Live View line exceeded the safety limit." }
            bytes.add(value.toByte())
        }
        if (bytes.lastOrNull() == '\r'.code.toByte()) bytes.removeAt(bytes.lastIndex)
        return String(bytes.toByteArray(), StandardCharsets.ISO_8859_1)
    }

    private enum class Boundary { NEXT, CLOSED }
}

internal class CcapiMultipartLiveViewSession(
    private val call: Call,
    private val response: Response,
    private val sourceUrl: String,
    boundary: String,
) : Closeable {
    private val monitor = Object()
    private var latestFrame: ByteArray? = null
    private var producedGeneration = 0L
    private var consumedGeneration = 0L
    private var terminalError: Throwable? = null
    private var closed = false
    private val worker = Thread({ drain(boundary) }, "ccapi-multipart-live-view").apply {
        isDaemon = true
        start()
    }

    fun nextFrame(timeoutSeconds: Long = 15): LiveViewFrame {
        val deadline = System.nanoTime() + TimeUnit.SECONDS.toNanos(timeoutSeconds)
        synchronized(monitor) {
            while (!closed && producedGeneration <= consumedGeneration && terminalError == null) {
                val remaining = deadline - System.nanoTime()
                if (remaining <= 0) error("Timed out waiting for the next Canon multipart Live View frame.")
                TimeUnit.NANOSECONDS.timedWait(monitor, remaining)
            }
            if (producedGeneration > consumedGeneration) {
                val frame = checkNotNull(latestFrame)
                consumedGeneration = producedGeneration
                return LiveViewFrame(frame.copyOf(), "image/jpeg", sourceUrl)
            }
            terminalError?.let { throw IllegalStateException("Canon multipart Live View stream failed.", it) }
            check(!closed) { "Canon multipart Live View stream is closed." }
            error("Canon multipart Live View did not produce a frame.")
        }
    }

    private fun drain(boundary: String) {
        try {
            val body = response.body ?: error("Canon multipart Live View returned an empty response body.")
            val reader = CcapiMultipartFrameReader(body.byteStream(), boundary)
            while (!closed) {
                val frame = reader.nextFrame() ?: error("Canon multipart Live View stream ended unexpectedly.")
                synchronized(monitor) {
                    latestFrame = frame
                    producedGeneration += 1
                    monitor.notifyAll()
                }
            }
        } catch (exception: Throwable) {
            synchronized(monitor) {
                if (!closed) terminalError = exception
                monitor.notifyAll()
            }
        } finally {
            response.close()
        }
    }

    override fun close() {
        synchronized(monitor) {
            if (closed) return
            closed = true
            monitor.notifyAll()
        }
        call.cancel()
        response.close()
        worker.interrupt()
    }
}
